packages/disputer/index.js

```diff
@@ -52,8 +52,13 @@ async function run({
     // Setup web3 accounts and network
     const [accounts, networkId] = await Promise.all([web3.eth.getAccounts(), web3.eth.net.getId()]);
 
-    // Setup contract instances. NOTE that getAddress("Voting", networkId) will resolve to null in tests.
-    const voting = new web3.eth.Contract(getAbi("Voting"), getAddress("Voting", networkId));
+    // Setup contract instances. NOTE getAddress("Voting", networkId) will error if run in tests. Catch & set it to null.
```

Comment:
Hmmm, this seems bad that we're not testing our bots' ability to grab our contracts from the artifacts.... I don't see a straightforward solution, though, other than to not rely on addresses in the artifacts at all (which seems like a nonstarter). Any ideas?

We could have node tests that run against infura that specifically test this, but `try-catching` this in the local tests doesn't seem to be scalable for infra that more heavily relies on global contracts.

One really nasty way of doing this is to do some sort of in-memory mocking or modification of the buidler artifacts. I think this is something we'll need a concrete solution for -- deploying local versions of global contracts at the start of the test and accessing them through artifacts will be a must-have I suspect.

Reply:
yup! I agree it's really not ideal. I was not super happy about this solution.

One solution is to modify the `getAddress` to return `null` as it does with truffle tests so this does not error out. We can then work on fixing this method to work correctly with hardhat by correctly deploying the voting contract within the tests and ensuring that this artifact is pulled in as expected.

The thing about this specific `voting` instance is that it is not tested within these `index.js` tests. these tests are testing the main entry point to ensure that the looping, retry and instantiation logic works. The actual testing of the `disputer.js` & `liquidator.js` tests are not affected by this as the voting contract's address is provided as a param in these tests.

Reply:
I’ve found a workaround for this. Hardhat’s internal structure is not the same as buidler and so we can't simply pull the address from the global return `global.artifacts._provisioner._deploymentAddresses[contractName];` the way we've been doing it before within the core package. However, a really simple workaround is to just set the addresses we need for this method within the global space within the tests and then they are accessible.

What I’ve done looks like this:
1. within the unit tests deploy the `Voting` contract.
2. within unit tests set `global.hardhatTestingAddresses["Voting"] = voting.address`
3. Change the `getAddressTest` to try to pull this address return `global.hardhatTestingAddresses[contractName]` before erroring out. 

This lets us remove the null logic checks within the bots, which is great.

One small pain point with this is that the bots (liquidator, disputer and monitor) are dependent on `"@uma/core": "npm:@uma/core@1.2.0"`. To test this logic I removed the `npm:@uma/core` part of the dependency so I could use my local package. However, this wont work with the PR as we still need the bots to be dependent on this old package until we’ve updated the bots to work with the new contracts. What we can do therefore is create a new core release that has the old contract interfaces & methods from `1.2.0` but includes the new logic for `hardhatTestingAddresses` within the `getAddressTest` method. This will, however, require a separate PR which includes this 1.2.1 release.